Core.Collectors.Tests/Web/FixedHttpClient.cs

```diff
@@ -39,25 +42,38 @@ public void AddResponseGenerator(string requestUrl, string responseBody, Func<Tu
             this.requestToResponseGeneratorMap.Add(requestUrl + responseBody, resposneGenerator);
         }
 
-        public void AddResponse(string requestUrl, HttpStatusCode responseStatusCode, string responseMessage)
+        public void AddResponse(string requestUrl, HttpStatusCode responseStatusCode, string responseMessage, Dictionary<string, List<string>> responseHeaders = null)
         {
-            this.requestToResponseMap.Add(requestUrl, Tuple.Create(responseStatusCode, responseMessage));
+            AddResponse(requestUrl, string.Empty, responseStatusCode, responseMessage, responseHeaders);
         }
 
-        public void AddResponse(string requestUrl, string requestBody, HttpStatusCode responseStatusCode, string responseMessage)
+        public void AddResponse(string requestUrl, string requestBody, HttpStatusCode responseStatusCode, string responseMessage, Dictionary<string, List<string>> responseHeaders = null)
         {
-            this.requestToResponseMap.Add(requestUrl + requestBody, Tuple.Create(responseStatusCode, responseMessage));
+            requestUrl += requestBody;
+            HttpResponseMessage response = new HttpResponseMessage()
+            {
+                StatusCode = responseStatusCode,
+                Content = new StringContent(responseMessage)
+            };
+
+            if (null != responseHeaders)
+            {
+                foreach (KeyValuePair<string, List<string>> header in responseHeaders)
+                {
+                    response.Headers.Add(header.Key, header.Value);
+                }
+            }
+
+            this.responseMap.Add(requestUrl, response);
         }
 
         public Task<HttpResponseMessage> GetAsync(string requestUrl, IAuthentication authentication)
         {
-            if (this.requestToResponseMap.TryGetValue(requestUrl, out Tuple<HttpStatusCode, string> response))
+            this.requestCount++;
+
+            if (this.responseMap.TryGetValue(requestUrl, out HttpResponseMessage response))
```

Comment:
Interesting, I could swear that I had to create a new HttpResponseMessage here since otherwise the Content stream would be exhausted (and becomes unreadable) if you make the same request back to back.

Can you create a small test (I thought I had this), that calls the same request twice and reads the content? If this works, fine, if this is throwing an exception, complaining about stream closed, etc., please re-create the HttpResponse here using the existing data in the map (you might need to be careful on how you read the content as reading it once again might cause stream to close).

Reply:
Looked into this and found that it is safe to read content multiple times as a string, but unsafe to read it as a stream. I have added a private method `CopyHttpResponseMessage` which creates a new object based using ReadAsStringAsync(). This ensures that object passed out of FixedHttpClient are always safe to read as either string or streams.